docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst

```diff
@@ -0,0 +1,421 @@
+    int data;
+    int x;
+  };
+
+  void process_user_data(int data) {
+    // ...
+    use_pointer(&data);
+  }
+
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->data);
+    free(u);
+  }
+
+Since data is no longer a pointer our ``isAdditionalFlowStep`` doesn't make any sense because it specifies flow to the indirection of the field (and an integer does not have any indirections). So there is no choice about whether to taint the value of the field or its indirection: it has to be the value. However, since we pass the address of ``data`` to ``use_pointer`` the tainted data is what is pointed to by the argument of ``use_pointer`` (since the data pointed to by ``&data`` is exactly ``data``). So to handle this case we need a mix of the two situations above:
+  1. We need to taint the value of the field just like in the :ref:`Using asExpr <using-asExpr>` section.
+  2. We need to select the indirection of the argument just like in the :ref:`Using asIndirectExpr <using-asIndirectExpr>` section.
+
+With these changes the query looks like:
+
+.. code-block:: ql
+
+  /**
+  * @kind path-problem
+  */
+
+  import semmle.code.cpp.dataflow.new.DataFlow
+  import Flow::PathGraph
+
+  module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) {
+      exists(Call call |
+        call.getTarget().hasName("write_user_input_to") and
+        source.asDefiningArgument() = call.getArgument(0)
+      )
+    }
+
+    predicate isSink(DataFlow::Node sink) {
+      exists(Call call |
+        call.getTarget().hasName("use_pointer") and
+        sink.asIndirectExpr() = call.getAnArgument()
+      )
+    }
+
+    predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
+      exists(FieldAccess fa |
+        n1.asIndirectExpr() = fa.getQualifier() and
+        n2.asExpr() = fa
+      )
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+And with that query the flow is identified.
+
+Specifying implicit reads
+-------------------------
+
+The previous section demonstrated how to add flow from qualifiers to field accesses because a source implicitly tainted all the fields of a struct. This section considers the opposite scenario: A specific field is being tainted, and we want to find any place that may read from this object, including any place that reads an unknown set of fields.
+
+To set the stage, consider the following scenario:
+
+.. code-block:: cpp
+
+  struct A {
+    const int *p;
+    int x;
+  };
+
+  struct B {
+    A *a;
+    int z;
+  };
+
+  int user_input();
+  void read_data(const void *);
+  void *malloc(size_t);
+
+  void get_input_and_use_data() {
+    B b;
+    b.a = (A *)malloc(sizeof(A));
+    b.a->x = user_input();
+    // ...
+    use_data(&b);
```

Comment:
```suggestion
    read_data(&b);
```
I think `use_data` and `read_data` are references to the same thing?

Reply:
Oops. Indeed they are. I had to fix a few other references, so I did it in c5307bed2b3c691c793b19218afe32b26b96d191 instead of accepting your commit suggestion.